EssentialApp/EssentialAppTests/ImageCommentsUIIntegrationTests.swift

```diff
@@ -0,0 +1,172 @@
+		assertThat(sut, isRendering: imageComments.presentableComments)
+		
+		sut.simulateUserInitiatedImageCommentsReload()
+		loader.completeCommentsLoadingWithError(at: 1)
+		assertThat(sut, isRendering: imageComments.presentableComments)
+	}
+	
+	func test_loadCommentsCompletion_rendersErrorMessageOnErrorUntilNextReload() {
+		let (sut, loader) = makeSUT()
+		
+		sut.loadViewIfNeeded()
+		XCTAssertEqual(sut.errorMessage, nil)
+		
+		loader.completeCommentsLoadingWithError(at: 0)
+		XCTAssertEqual(sut.errorMessage, localized("IMAGE_COMMENTS_VIEW_CONNECTION_ERROR"))
+		
+		sut.simulateUserInitiatedImageCommentsReload()
+		XCTAssertEqual(sut.errorMessage, nil)
+	}
+	
+	func test_tapOnErrorView_hidesErrorMessage() {
+		let (sut, loader) = makeSUT()
+		
+		sut.loadViewIfNeeded()
+		XCTAssertEqual(sut.errorMessage, nil)
+		
+		loader.completeCommentsLoadingWithError(at: 0)
+		XCTAssertEqual(sut.errorMessage, localized("IMAGE_COMMENTS_VIEW_CONNECTION_ERROR"))
+		
+		sut.simulateErrorViewTap()
+		XCTAssertEqual(sut.errorMessage, nil)
+	}
+	
+	func test_deinit_cancelsRunningRequest() {
+		var sut: ImageCommentsViewController?
+		let loader = LoaderSpy()
+		
+		autoreleasepool {
+			sut = ImageCommentsUIComposer.imageCommentsComposedWith(commentsLoader: loader)
+			
+			sut?.loadViewIfNeeded()
+		}
+		
+		XCTAssertEqual(loader.cancelCallCount, 0, "Expected no cancelled requests until comments are not visibile")
+		
+		sut = nil
+		
+		XCTAssertEqual(loader.cancelCallCount, 1, "Expected one cancelled request")
+	}
+	
+	func test_loadCommentsCompletion_dispatchesFromBackgroundToMainThread() {
+		let (sut, loader) = makeSUT()
+		sut.loadViewIfNeeded()
+		
+		let exp = expectation(description: "Wait for background queue")
+		DispatchQueue.global().async {
+			loader.completeCommentsLoading(at: 0)
+			exp.fulfill()
+		}
+		wait(for: [exp], timeout: 1.0)
+	}
+		
+	// MARK: - Helpers
+	
+	private func makeSUT(file: StaticString = #filePath, line: UInt = #line) -> (sut: ImageCommentsViewController, loader: LoaderSpy) {
+		let loader = LoaderSpy()
+		let sut = ImageCommentsUIComposer.imageCommentsComposedWith(commentsLoader: loader)
+		trackForMemoryLeaks(loader, file: file, line: line)
+		trackForMemoryLeaks(sut, file: file, line: line)
+		return (sut, loader)
+	}
+	
+	private func uniqueComments(currentDate: Date = Date()) -> (comments: [ImageComment], presentableComments: [PresentableImageComment]) {
+		let comments = [
+			ImageComment(id: UUID(),
+						 message: "a message",
+						 createdAt: Date(timeInterval: -dayInSeconds, since: currentDate),
+						 author: ImageCommentAuthor(username: "a username")),
+			ImageComment(id: UUID(),
+						 message: "another message",
+						 createdAt: Date(timeInterval: -hourInSeconds, since: currentDate),
+						 author: ImageCommentAuthor(username: "another username"))]
+		
+		let presentableComments = ImageCommentsPresenter.map(comments, currentDate: { currentDate })
```

Comment:
The `currentDate` is irrelevant in this test so you don't need to pass it.

```suggestion
		let presentableComments = ImageCommentsPresenter.map(comments)
```